src/sage/schemes/elliptic_curves/ell_finite_field.py

```diff
@@ -1341,6 +1343,96 @@ def is_ordinary(self, proof=True):
         """
         return not is_j_supersingular(self.j_invariant(), proof=proof)
 
+    def has_order(self, value, num_checks=8):
+        r"""
+        Return ``True`` if the curve has order ``value``.
+
+        INPUT:
+
+        - ``value`` -- integer in the Hasse-Weil range for this curve
+
+        - ``num_checks``-- integer (default: `8`); the number of times to check
+          whether ``value`` times a random point on this curve equals the
+          identity
+
+        .. NOTE::
+
+            Since the method is probabilistic, there is a possibility for the
+            method to yield false positives (i.e. returning ``True`` even when
+            the result is ``False``). Even worse, it is possible for this to
+            happen even when ``num_checks`` is increased arbitrarily. See below
+            for an example and :issue:`38617` for an open discussion.
+
+        EXAMPLES:
+
+        For curves over small finite fields, the order is computed and compared
+        directly::
+
+            sage: E = EllipticCurve(GF(7), [0, 1])
+            sage: E.order()
+            12
+            sage: E.has_order(12, num_checks=0)
+            True
+            sage: E.has_order(11, num_checks=0)
+            False
+            sage: E.has_order(13, num_checks=0)
+            False
+
+        This tests the method on a random curve::
+
+            sage: # long time (10s)
+            sage: p = random_prime(2**128, lbound=2**127)
+            sage: K = GF((p, 2), name="a")
+            sage: E = EllipticCurve(K, [K.random_element() for _ in range(2)])
+            sage: N = E.order()
+            sage: E.has_order(N, num_checks=20)
+            True
+            sage: E.has_order(N + 1)
+            False
```

Comment:
Basically we introduced `has_order` as a way to check a candisate curve has the right prime order `N`.
The method we used before was naive and much more costly: compute order of the curve and compare to `N`.
Now we want to check that the curve has the right order so we sample random points and check that they are of order `N` which is much easier to do (only a few double and add). The number of samples can be reduced to 1 since N is prime and we use `set_order` after this and it's supposed to yield a deterministic result.

Reply:
How is $p \sim 2^{128}$ not "big enough"?

Reply:
You're right, not sure what I was thinking. Maybe I misread this part. I still have a question about this example.

Since this algorithm sometimes fails and this test is random, are we potentially introducing a test that might randomly fail? That could be annoying for the CI. I'd prefer something like `p = nextprime(2^127)` or something than a test that might randomly fail.

Reply:
Nevermind, I think I misunderstood the failure example. This `has_order` can give false positives but never false negatives, is my understanding correct?

Reply:
Your understanding is correct, but it's not why this test doesn't fail. For that you have to understand the bug analysis in # 38617. To summarise it, for `E.has_order(N)` to pass, you need $\mathrm{LCM}(\mathrm{ord}(P)) \mid N$ over all $P$ sampled in the random test. Here since $E$ has ("real") prime order $N$, the LCM would also be $N$ (because $P \neq O_E$ in `has_order`, patched in this PR as well). But $N \mid N + 1$ never holds for prime $N$ so the test never fails.